File: JavaScript/17_Binary Tree-I/M-TUF-Bottom view of Binary Tree.js
// TUF: Bottom View of Binary Tree
// https://takeuforward.org/plus/dsa/problems/bottom-view-of-bt

/**
 * Creates a tree node
 */
function createTreeNode(val = 0, left = null, right = null) {
    return { val, left, right };
}

/**
 * Get bottom view of binary tree
 * @param {TreeNode} root - Root of the binary tree
 * @returns {number[]} - Bottom view from left to right
 */
function bottomView(root) {
    if (root === null) return [];

    const map = new Map(); // horizontal distance -> node value
    const queue = [[root, 0]]; // [node, hd]

    while (queue.length > 0) {
        const [node, hd] = queue.shift();

        // For bottom view, overwrite value at HD every time
        map.set(hd, node.val);

        if (node.left !== null) {
            queue.push([node.left, hd - 1]);
        }
        if (node.right !== null) {
            queue.push([node.right, hd + 1]);
        }
    }

    // Sort by horizontal distance and get values
    const sortedHDs = [...map.keys()].sort((a, b) => a - b);
    return sortedHDs.map(hd => map.get(hd));
}

// Test cases
const root = createTreeNode(1);
root.left = createTreeNode(2);
root.right = createTreeNode(3);
root.left.right = createTreeNode(5);
root.right.right = createTreeNode(4);

console.log("Bottom View:", bottomView(root)); // [2, 5, 3, 4]

const root2 = createTreeNode(20);
root2.left = createTreeNode(8);
root2.right = createTreeNode(22);
root2.left.left = createTreeNode(5);
root2.left.right = createTreeNode(3);
root2.right.right = createTreeNode(25);
root2.left.right.left = createTreeNode(10);
root2.left.right.right = createTreeNode(14);

console.log("Bottom View 2:", bottomView(root2)); // [5, 10, 3, 14, 25]

File: Java/17_Binary Tree-I/M-TUF-Bottom view of Binary Tree.java
// TUF: Bottom view of Binary Tree
// https://takeuforward.org/plus/dsa/problems/bottom-view-of-bt

import java.util.*;

class Solution {
    public static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode(int val) { this.val = val; }
    }

    // Bottom View
    public List<Integer> bottomView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;

        Map<Integer, Integer> map = new TreeMap<>(); // HD -> node.val
        Queue<Pair> queue = new LinkedList<>();
        queue.offer(new Pair(root, 0));

        while (!queue.isEmpty()) {
            Pair p = queue.poll();
            TreeNode node = p.node;
            int hd = p.hd;

            // For bottom view, overwrite value at HD every time
            map.put(hd, node.val);

            if (node.left != null) queue.offer(new Pair(node.left, hd - 1));
            if (node.right != null) queue.offer(new Pair(node.right, hd + 1));
        }

        for (int val : map.values()) {
            result.add(val);
        }
        return result;
    }

    // Helper class to store node and horizontal distance
    private static class Pair {
        TreeNode node;
        int hd;
        Pair(TreeNode n, int h) { node = n; hd = h; }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // Build tree: [1, 2, 3, null, 5, null, 4]
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.right = new TreeNode(5);
        root.right.right = new TreeNode(4);

        System.out.println("Bottom View: " + sol.bottomView(root)); // [2, 5, 3, 4]
    }
}
